shared_python/FinalTables.py

```diff
@@ -1,80 +1,118 @@
+        self.log = log
+
+    def original_table(self, table_name, filter='', database_name=None):
+        if table_name is None:
+            return None
+        if database_name is None:
+            original_database = self.original_database
+        else:
+            original_database = database_name
+        query = "SELECT * FROM `{0}`.`{1}` {2}".format(original_database, table_name, filter)
+        self.dict_cursor.execute(query)
+        return self.dict_cursor.fetchall()
+
+    def _escape_unescape(self, item):
+        return html.unescape(item).replace('\\', '\\\\').replace('"', '\\"').replace("'", "\\'")
+
+    def _value(self, row):
+        value = []
+        for item in row:
+            if type(item) is str:
+                value.append('"' + self._escape_unescape(item) + '"')
+            elif type(item) is datetime.datetime:
+                value.append('"' + str(item) + '"')
+            elif item is None:
+                value.append('null')
+            elif item == '':
+                value.append('""')
+            else:
+                value.append(str(item))
+        return value
+
+    def insert_into_final(self, output_table_name, rows, target_database=None):
+        if target_database:
+            final_database = target_database
+        else:
+            final_database = self.final_database
+        self.dict_cursor.execute("TRUNCATE `{0}`.`{1}`".format(final_database, output_table_name))
+        columns = rows[0].keys()
+        values = []
+        for row in rows:
+            col = self._value(row.values())
+            values.append('(' + ', '.join(col) + ')')
+
+        self.dict_cursor.execute(f"""
+            INSERT INTO `{final_database}`.`{output_table_name}` ({', '.join(columns)}) VALUES {', '.join(values)}
+          """)
+        self.db.commit()
+
+    def populate_story_tags(self, story_id, output_table_name, story_tags):
+        cols_with_tags = []
+        for (col, tags) in story_tags.items():
+            cols_with_tags.append(u"{0}='{1}'".format(col, tags.replace("'", "\\'").strip()))
+
+        if cols_with_tags:
+            self.dict_cursor.execute("""
          UPDATE `{0}`.`{1}` SET {2} WHERE id={3}
         """.format(self.final_database, output_table_name, ", ".join(cols_with_tags), story_id))
-      self.db.commit()
+            self.db.commit()
+
+    def story_to_final_without_tags(self, story, story_authors, is_story=True):
+        type = 'story' if is_story else 'story_link'
+        authors_count = len(story_authors)
+        notes = story['notes']
+        if authors_count > 2:
+            # AO3 works can't currently be imported with more than two authors
+            self.log.warning(f"{type} {story['id']} has {authors_count} authors - listing all authors in notes...")
+            author_names = "Creators: {} and {}".format(", ".join(story_authors[:-1]),  story_authors[-1])
+            notes = "{author_names}<br/><br/>{notes}" if notes else author_names
```

Comment:
Do we want to include some text here saying "Authored by" or similar?

Reply:
I went for "Creators:" because it's consistent with the AO3 terminology. I didn't want to get too fancy as the text will have to be hard-coded in English which might be confusing/annoying if the rest of the work is actually in a different language. But maybe I should really run that past the chairs in case there's a better option!